### PR TITLE
[common/protected_files] Add support for arbitrary truncation

### LINKED ISSUE #950: [LibOS] Add support for shrinking encrypted files via truncate(2) to arbitrary size
Graminizing the ethereum [geth client](https://github.com/ethereum/go-ethereum), we are running into another issue regarding encrypted files. To protect the state of geth's persistent data from host access and manipulation, files are stored in a encrypted mount point. 

Syncing `goerli` testnet with geth into encrypted mount point on Gramine 1.3.1, we run into this error:
```
[P1:T12:geth] warning: encrypted_file_set_size: pf_set_size failed: Functionality not implemented
CRIT [10-03|17:28:11.357] Failed to truncate ancient data          number=1,325,056 err="truncate /root/.ethereum/goerli/geth/chaindata/ancient/chain/bodies.cidx: permission denied"
```
At times, geth needs to rewind the state of the blockchain to a previous block. It does so by shrinking the size of the database file to the specific block. Since shrinking encrypted files is not supported within gramine, this operation fails.

### PR DESCRIPTION
**Pull request by mwkmwkmwk**
## Description of the changes

Fixes #950.

Closes #972.

## How to test this PR?

Tests included in PR.

- - -
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)

### CODE REVIEW

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 8 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

___
*[`common/src/protected_files/protected_files.c` line 1311 at r1](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdkpQz1aBU7vRSzYXz:-NrfVN9J7y6Sn77A5AQ9:b-5kudg9) ([raw file](https://github.com/gramineproject/gramine/blob/8f6546dea9914ceff29a1acb6e7812ae391e3f61/common/src/protected_files/protected_files.c#L1311)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I'm confused what this flag actually does. Why do we need writing, if we just truncated (i.e. discarded the some end-part of the file)? We already flushed everything that was pending for write at the very beginning.

At the same time, this seems to not hurt, so I'm not blocking.
</blockquote></details>

this is needed because we change the metadata in the first block (ie. new size)
___
*[`common/src/protected_files/protected_files.c` line 1312 at r1](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdkR-68VuoU5YHWpX2:-NrfVVOZ5Ya7Zb0JaC19:b-g3wi5x) ([raw file](https://github.com/gramineproject/gramine/blob/8f6546dea9914ceff29a1acb6e7812ae391e3f61/common/src/protected_files/protected_files.c#L1312)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Btw, `real_file_size` is not used at all in Protected Files, just noticed that. So this can be safely removed.

Or at least make it uniform -- in the above case of `truncate to 0`, this size is set to zero, which is incorrect -- it must be at least one (root MHT) node size.
</blockquote></details>

yeah, I noticed that too, I just wanted to avoid unrelated changes in the same PR

(this isn't the only unused field btw)

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 8 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdfXF15YO6wBnbKdA-:-NriLCSd61ygt85BB74B:b-896fix) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Fix this: https://github.com/gramineproject/gramine/blob/63ee38d5eb28f5a880a973c28de7be6f0dc6f290/common/src/protected_files/README.rst?plain=1#L69
</blockquote></details>

Done.

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1785#-Nrde_0M4Eo1fyYPVofh:-NriLD-W72JklZk5FAZM:b-896fix) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Please fix our Documentation: https://github.com/gramineproject/gramine/blob/63ee38d5eb28f5a880a973c28de7be6f0dc6f290/Documentation/devel/features.md?plain=1#L2090-L2092
</blockquote></details>

Done.

___
___
*[`common/src/protected_files/protected_files.c` line 1286 at r1](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdhkMX8NN34Q762Q8S:-NriLMnK3nj2heyV3jZT:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/8f6546dea9914ceff29a1acb6e7812ae391e3f61/common/src/protected_files/protected_files.c#L1286)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

But "simply flushing" means `ipf_internal_flush()`. Your sentence is a bit confusing: you want to flush the data first, so that you can empty the cache. Could you rephrase?
</blockquote></details>

Done.
___
*[`libos/test/fs/seek_tell_truncate.c` line 5 at r1](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdZOC40l6TELZteQwm:-NriLVzO5r4CIb939qMu:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/8f6546dea9914ceff29a1acb6e7812ae391e3f61/libos/test/fs/seek_tell_truncate.c#L5)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`void* `
</blockquote></details>

Done.

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: 1 of 5 files reviewed, 8 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

___
*[`common/src/protected_files/protected_files.c` line 1294 at r1](https://reviewable.io/reviews/gramineproject/gramine/1785#-Nrdg274CWGfXsY4klG5:-NriMhtOAl7-WS1N2DH3:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/8f6546dea9914ceff29a1acb6e7812ae391e3f61/common/src/protected_files/protected_files.c#L1294)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

This "delete completely" snippet we already have 3 times in the current codebase, and you add one more. Can we introduce a helper function?
</blockquote></details>

Done.

**Review by dimakuv — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: 1 of 5 files reviewed, 8 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @mwkmwkmwk)

___
*[`-- commits` line 8 at r3](https://reviewable.io/reviews/gramineproject/gramine/1785#-NriU0QSAw4sCPPFC0ZV):*
Just FYI: It's better to add the Signed-off-by line even in fixup commits. This makes the GitHub DCO plugin happy (otherwise it's red and makes an impression your PR is failing our CI).

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: 1 of 5 files reviewed, 8 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

___
*[`-- commits` line 8 at r3](https://reviewable.io/reviews/gramineproject/gramine/1785#-NriU0QSAw4sCPPFC0ZV):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Just FYI: It's better to add the Signed-off-by line even in fixup commits. This makes the GitHub DCO plugin happy (otherwise it's red and makes an impression your PR is failing our CI).
</blockquote></details>

I see, very well then

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @mwkmwkmwk)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdemFG9GUiRZnAhXPz:-NriV8H58tfwNHzUt2hA:b-875zyl) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

What about this overridden test: https://github.com/gramineproject/gramine/blob/63ee38d5eb28f5a880a973c28de7be6f0dc6f290/libos/test/fs/test_enc.py#L162-L164

I mean, you need to:
- fix the comment
- add more cases in `test_140_file_truncate` for Encrypted Files (in the same file that you modified in this PR)
</blockquote></details>

Not yet done?

___
___
*[`common/src/protected_files/protected_files.c` line 1311 at r1](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdkpQz1aBU7vRSzYXz:-NriUpuo1Wa_xh90ZntH:b-gxztsp) ([raw file](https://github.com/gramineproject/gramine/blob/8f6546dea9914ceff29a1acb6e7812ae391e3f61/common/src/protected_files/protected_files.c#L1311)):*
<details><summary><i>Previously, mwkmwkmwk (Marcelina Kościelnicka) wrote…</i></summary><blockquote>

this is needed because we change the metadata in the first block (ie. new size)
</blockquote></details>

Fair enough.
___
*[`common/src/protected_files/protected_files.c` line 1312 at r1](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdkR-68VuoU5YHWpX2:-NriUsfT214286i0LSwJ:b-e932cy) ([raw file](https://github.com/gramineproject/gramine/blob/8f6546dea9914ceff29a1acb6e7812ae391e3f61/common/src/protected_files/protected_files.c#L1312)):*
<details><summary><i>Previously, mwkmwkmwk (Marcelina Kościelnicka) wrote…</i></summary><blockquote>

yeah, I noticed that too, I just wanted to avoid unrelated changes in the same PR

(this isn't the only unused field btw)
</blockquote></details>

Do you plan to submit a separate PR fixing this? Should be trivial.

Alternatively, I can do it, but you need to tell me what other unused fields are there.

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners

___
*[`common/src/protected_files/protected_files.c` line 1312 at r1](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdkR-68VuoU5YHWpX2:-NriYgPlAAJy0qCnObkW:b41rdty) ([raw file](https://github.com/gramineproject/gramine/blob/8f6546dea9914ceff29a1acb6e7812ae391e3f61/common/src/protected_files/protected_files.c#L1312)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Do you plan to submit a separate PR fixing this? Should be trivial.

Alternatively, I can do it, but you need to tell me what other unused fields are there.
</blockquote></details>

Done in #1789.

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdemFG9GUiRZnAhXPz:-NriYnfh3433iS4XMgAq:b-vbtecn) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Not yet done?
</blockquote></details>

Yeah, still working on it

___

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: 4 of 5 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

___
*[`common/src/protected_files/protected_files.c` line 1312 at r1](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdkR-68VuoU5YHWpX2:-Ns8pwWO7xZuF5yKEfMn:b293gog) ([raw file](https://github.com/gramineproject/gramine/blob/8f6546dea9914ceff29a1acb6e7812ae391e3f61/common/src/protected_files/protected_files.c#L1312)):*
<details><summary><i>Previously, mwkmwkmwk (Marcelina Kościelnicka) wrote…</i></summary><blockquote>

Done in #1789.
</blockquote></details>

Rebased the PR on top.

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: 4 of 6 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1785#-NrdemFG9GUiRZnAhXPz:-Ns8qnGtC5MmWvBwzARZ:b-3b5zip) (no related file):*
<details><summary><i>Previously, mwkmwkmwk (Marcelina Kościelnicka) wrote…</i></summary><blockquote>

Yeah, still working on it
</blockquote></details>

I ended up just removing the override, since the base tests seem alright already.

___

**Review by dimakuv — DISMISSED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel)

**Review by kailun-qin — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mwkmwkmwk)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1785#-NsBGNEtCkx4uOE8sb1A:-NsBGNEtCkx4uOE8sb1B:b-2mliu6) (no related file):*
I think we also need to update https://github.com/gramineproject/gramine/blob/4afc5509cdd7a91b391f83407860f5607c323600/common/src/protected_files/protected_files.h#L284-L285 and https://github.com/gramineproject/gramine/blob/4afc5509cdd7a91b391f83407860f5607c323600/libos/src/fs/chroot/encrypted.c#L28-L29.

___

**Review by kailun-qin — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mwkmwkmwk)

___
*[`Documentation/devel/features.md` line 2166 at r4](https://reviewable.io/reviews/gramineproject/gramine/1785#-NsBKd5wDsAWctbaW4x0:-NsBKd5wDsAWctbaW4x1:bdvw8ud) ([raw file](https://github.com/gramineproject/gramine/blob/b11f67301e19d4f69a647f0a634df7ecceeba753/Documentation/devel/features.md#L2166)):*
> ```Markdown
> - ☑ `fdatasync()`
> - ▣ `truncate()`: see note above
> - ▣ `ftruncate()`: see note above
> ```

What about here?



_Code quote:_
```Markdown
- ▣ `truncate()`: see note above
- ▣ `ftruncate()`: see note above
```

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @kailun-qin)

___
*[`Documentation/devel/features.md` line 2166 at r4](https://reviewable.io/reviews/gramineproject/gramine/1785#-NsBKd5wDsAWctbaW4x0:-NsI0r6X6lwsuUoG2xBI:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/b11f67301e19d4f69a647f0a634df7ecceeba753/Documentation/devel/features.md#L2166)):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

What about here?
</blockquote></details>

Done.

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: 5 of 6 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @kailun-qin)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1785#-NsBGNEtCkx4uOE8sb1A:-NsI17OVBf7P7x4rounW:b-896fix) (no related file):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

I think we also need to update https://github.com/gramineproject/gramine/blob/4afc5509cdd7a91b391f83407860f5607c323600/common/src/protected_files/protected_files.h#L284-L285 and https://github.com/gramineproject/gramine/blob/4afc5509cdd7a91b391f83407860f5607c323600/libos/src/fs/chroot/encrypted.c#L28-L29.
</blockquote></details>

Done.

___

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @kailun-qin and @mwkmwkmwk)

___
*[`Documentation/devel/features.md` line 2166 at r4](https://reviewable.io/reviews/gramineproject/gramine/1785#-NsBKd5wDsAWctbaW4x0:-NsIFx-66D3Kv8SrmH2w:bi6h179) ([raw file](https://github.com/gramineproject/gramine/blob/b11f67301e19d4f69a647f0a634df7ecceeba753/Documentation/devel/features.md#L2166)):*
<details><summary><i>Previously, mwkmwkmwk (Marcelina Kościelnicka) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

There are no "notes above" anymore, please remove this part (2x)
___
*[`libos/src/fs/chroot/encrypted.c` line 28 at r5](https://reviewable.io/reviews/gramineproject/gramine/1785#-NsIG1fb977ldki8V9ZW:-NsIG1fb977ldki8V9ZX:b-2qep2y) ([raw file](https://github.com/gramineproject/gramine/blob/fd7cf63dfdcd487bb74d5e1ab73dd92d446120bc/libos/src/fs/chroot/encrypted.c#L28)):*
> ```C
>  * TODO:
>  *
>  * - flush all files on process exit
> ```

I have a strong suspicion that's implemented for some time now. Can you check?

**Review by dimakuv — COMMENTED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @kailun-qin and @mwkmwkmwk)

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @kailun-qin and @mwkmwkmwk)

___
*[`libos/src/fs/chroot/encrypted.c` line 28 at r5](https://reviewable.io/reviews/gramineproject/gramine/1785#-NsIG1fb977ldki8V9ZW:-NsSM8S63tye-4vaSMb6:b-3u4kz5) ([raw file](https://github.com/gramineproject/gramine/blob/fd7cf63dfdcd487bb74d5e1ab73dd92d446120bc/libos/src/fs/chroot/encrypted.c#L28)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I have a strong suspicion that's implemented for some time now. Can you check?
</blockquote></details>

I'm actually not sure what this TODO could mean. At process exit, libc (or whatever runtime) is supposed to `close()` all opened files. Then, a `chroot_encrypted_close` of each encrypted file will end up in here: 

https://github.com/gramineproject/gramine/blob/d82e885cb4adbcec876c78a0f745559cdb26aee4/common/src/protected_files/protected_files.c#L1126-L1134

So each encrypted file will be flushed.

In other words, I think this is already done. I would remove this TODO.

**Review by mkow — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @kailun-qin)

___
*[`libos/src/fs/chroot/encrypted.c` line 28 at r5](https://reviewable.io/reviews/gramineproject/gramine/1785#-NsIG1fb977ldki8V9ZW:-NsTJMx9Er_bsc-B8Pdb:ba8qtst) ([raw file](https://github.com/gramineproject/gramine/blob/fd7cf63dfdcd487bb74d5e1ab73dd92d446120bc/libos/src/fs/chroot/encrypted.c#L28)):*
I'm sure it wasn't the case, at least some time ago.

> At process exit, libc (or whatever runtime) is supposed to `close()` all opened files.

But that's a different thing? We're on the "kernel" level, the process may exit without going through libc, it may also be using a different runtime than glibc. It should be Gramine who flushes buffers on process exit when destroying handles.

**Review by dimakuv — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @kailun-qin and @mkow)

___
*[`libos/src/fs/chroot/encrypted.c` line 28 at r5](https://reviewable.io/reviews/gramineproject/gramine/1785#-NsIG1fb977ldki8V9ZW:-NsUDRTjCF1lcUHISJQn:b-60u0td) ([raw file](https://github.com/gramineproject/gramine/blob/fd7cf63dfdcd487bb74d5e1ab73dd92d446120bc/libos/src/fs/chroot/encrypted.c#L28)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

I'm sure it wasn't the case, at least some time ago.

> At process exit, libc (or whatever runtime) is supposed to `close()` all opened files.

But that's a different thing? We're on the "kernel" level, the process may exit without going through libc, it may also be using a different runtime than glibc. It should be Gramine who flushes buffers on process exit when destroying handles.
</blockquote></details>

Ok, I see what you mean.

But looking again, I see that Gramine closes all handles on process exit (if I read the code right):
- https://github.com/gramineproject/gramine/blob/d82e885cb4adbcec876c78a0f745559cdb26aee4/libos/src/sys/libos_exit.c#L140
- https://github.com/gramineproject/gramine/blob/d82e885cb4adbcec876c78a0f745559cdb26aee4/libos/src/bookkeep/libos_handle.c#L713-L721
- https://github.com/gramineproject/gramine/blob/d82e885cb4adbcec876c78a0f745559cdb26aee4/libos/src/bookkeep/libos_handle.c#L707-L711

Note `put_handle()`. However, I'm not sure if all handles will actually be completely closed...

Anyway, there is no explicit flush invocation, and this comment is unrelated anyway, so I'm unblocking.

**Review by kailun-qin — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 2 of 4 files at r2, 2 of 2 files at r4, 2 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, 1 unresolved discussion, not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @mkow)

**Review by mwkmwkmwk — COMMENTED** [review]
Removed special case for `size == 0` (asked by mkow)

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785#-:-NujpIt5C12vO88uld2q:b-ivyby8)* status: 7 of 8 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv, @kailun-qin, and @mkow)

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 2 of 4 files at r2, 1 of 2 files at r4, 2 of 3 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners

**Review by mwkmwkmwk — COMMENTED** [review]
Rebased on current master

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785#-:-NukQJbF1GAtWdRunMox:b-luhjwb)* status: 7 of 8 files reviewed, all discussions resolved (waiting on @dimakuv, @kailun-qin, and @mkow)

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

**Review by kailun-qin — APPROVED** [review]
> Reviewed 1 of 1 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1785)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved
